src/iop/filmicrgb.c

```diff
@@ -423,117 +493,905 @@ static inline float linear_saturation(const float x, const float luminance, cons
+   *    by a sigmoid centered in `reconstruct_threshold` so the transition is soft and symmetrical
+   */
+
+  int clipped = 0;
+
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(in, mask, normalize, feathering, width, height, ch) \
+  schedule(simd:static) aligned(mask, in:64) reduction(+:clipped)
+#endif
+  for(size_t k = 0; k < height * width * ch; k += ch)
+  {
+    const float pix_max = sqrtf(sqf(in[k]) + sqf(in[k+ 1]) + sqf(in[k + 2]));
+    const float argument = -pix_max * normalize + feathering;
+    const float weight = 1.0f / ( 1.0f + exp2f(argument));
+    mask[k / ch] = weight;
+
+    // at x = 4, the sigmoid produces opacity = 5.882 %.
+    // any x > 4 will produce neglictible changes over the image,
+    // especially since we have reduced visual sensitivity in highlights.
+    // so we discard pixels for argument > 4. for they are not worth computing.
+    clipped += (4.f > argument);
+  }
+
+  // If clipped area is < 9 pixels, recovery is not worth the computational cost, so skip it.
+  return (clipped > 9);
+}
+
+
+// B spline filter
+#define fsize 5
+
+
+#ifdef _OPENMP
+#pragma omp declare simd aligned(in, out:64) uniform(mult, bound_left, bound_right, ch, width, height)
+#endif
+inline static void blur_2D_Bspline_vertical(const float *const restrict in, float *const restrict out,
+                                            const size_t width, const size_t height, const size_t ch, const size_t mult,
+                                            const int bound_left, const int bound_right)
+{
+  // À-trous B-spline interpolation/blur shifted by mult
+  // Convolve B-spline filter over lines
+  #ifdef _OPENMP
+  #pragma omp parallel for default(none) \
+    dt_omp_firstprivate(in, out, width, height, ch, bound_left, bound_right, mult) \
+    schedule(simd:static) collapse(2)
+  #endif
+  for(size_t i = 0; i < height; i++)
+    for(size_t j = 0; j < width; j++)
+    {
+      const size_t index_out = (i * width + j) * ch;
+      float DT_ALIGNED_PIXEL accumulator[4] = { 0.0f };
+
+      // Are we in the boundary zone that needs bound checking ?
+      const int check = !((j > 2 * mult) && (j < width - 2 * mult));
+
+      // -funswitch-loops should compile 2 loops, for each check outcome
+      if(check)
+      {
+        #ifdef _OPENMP
+        #pragma omp simd aligned(in:64) aligned(accumulator:16) reduction(+:accumulator)
+        #endif
+        for(size_t jj = 0; jj < fsize; ++jj)
+          for(size_t c = 0; c < 3; ++c)
+          {
+            int index_x = mult * (jj - (fsize - 1) / 2) + j;
+            index_x = (index_x < bound_left)  ? bound_left  :
+                      (index_x > bound_right) ? bound_right :
+                                                index_x     ;
+
+            static const float DT_ALIGNED_ARRAY filter[fsize] = { 1.0f / 16.0f, 4.0f / 16.0f, 6.0f / 16.0f, 4.0f / 16.0f, 1.0f / 16.0f };
```

Comment:
I had it out of the loop before, but it turned out to be slower since it needs to be initialized from the outside for each thread. Since it's static const, I guess OpenMP parallel loops can cache the array in the code and avoid some memory flip-flops at runtime.

Reply:
So maybe a macro to share the code for all values ?

Reply:
When it was outside, was it declared as static const too? Because IMHO it shouldn't matter where static const declaration is put, it may as well be a global variable - it's allocated in the same space and not initialized at runtime, but at compile time.

Reply:
Yes, it was static const both cases, but I had to declare is first_private anyway for OpenMP, so I suspect it is the cause. Using a macro seems fine.

Reply:
It doesn't look like we can define arrays in macros

Reply:
Have you tried making it shared instead of first_private then? I see that first_private is considered a better choice, though I'm not too sure why since I don't know OpenMP well enough.

Reply:
Shared is really bad: it means only one thread can access the variable at one time, so every thread needs to wait for its turn to lock it. It's what you want to avoid at all cost, since it usually destroys the whole point of using multithreading. The goal here is to make the filter 100% private in threads.

Reply:
So it's impossible with OpenMP to access global memory directly, even constant? OK then.

Reply:
I wouldn't say I fully understand what OpenMP does here, but it seems indeed that its threads don't care about global memory.